### PR TITLE
H-705: Enable Sentry integration in `hash-graph`

### PR DESCRIPTION
**Pull request by indietyp**
## 🌟 What is the purpose of this PR?

This enables Sentry integration into the hash-graph. Like OTLP, Sentry is essentially disabled if the `HASH_GRAPH_SENTRY_DSN` is unset.

To properly spawn sentry the configuration of the runtime has been delegated, as the sentry guard needs to be initialized before tokio starts (it even works if that's not the case, but then measurements can be a bit misleading).

### 🚢 Has this modified a publishable library?

This PR:

- [x] does not modify any publishable blocks or libraries, or modifications do not need publishing

### 📜 Does this require a change to the docs?

The changes in this PR:

- [x] are internal and do not require a docs change

### 🕸️ Does this require a change to the Turbo Graph?

The changes in this PR:

- [x] do not affect the execution graph

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -64,6 +64,7 @@ HASH_REDIS_HOST=localhost
 HASH_REDIS_PORT=6379
 
 HASH_OTLP_ENDPOINT=http://localhost:4317
+HASH_GRAPH_SENTRY_DSN=
```
Comment (TimDiekmann) [inline .env:67]: We should not set an empty default in `.env`

File: apps/hash-graph/bin/cli/src/args.rs [hunk 2]
```diff
@@ -6,6 +6,9 @@ use crate::subcommand::Subcommand;
 #[derive(Debug, Parser)]
 #[clap(version, author, about, long_about = None)]
 pub struct Args {
+    #[arg(long, env = "HASH_GRAPH_SENTRY_DSN")]
+    pub sentry_dsn: Option<Option<sentry::types::Dsn>>,
```
Comment (TimDiekmann) [inline apps/hash-graph/bin/cli/src/args.rs:10]: Why this is a nested `Option`?

File: apps/hash-graph/bin/cli/src/main.rs [hunk 3]
```diff
@@ -5,27 +5,33 @@ mod args;
...
-        Subcommand::Completions(ref args) => {
-            subcommand::completions(args);
-            Ok(())
-        }
-        Subcommand::Snapshot(args) => subcommand::snapshot(args).await,
-        #[cfg(all(hash_graph_test_environment, feature = "test-server"))]
-        Subcommand::TestServer(args) => subcommand::test_server(args).await,
-    }
+    let _sentry = sentry::init(sentry::ClientOptions {
+        dsn: args.sentry_dsn.clone().flatten(),
```
Comment (TimDiekmann) [inline apps/hash-graph/bin/cli/src/main.rs:21]: Why the `sentry_dsn` has to be cloned? You're destructuring `args`and only call `execute` on another variable below.

File: apps/hash-graph/bin/cli/src/main.rs [hunk 4]
```diff
@@ -5,27 +5,33 @@ mod args;
...
-        Subcommand::TypeFetcher(args) => subcommand::type_fetcher(args).await,
-        Subcommand::Completions(ref args) => {
-            subcommand::completions(args);
-            Ok(())
-        }
-        Subcommand::Snapshot(args) => subcommand::snapshot(args).await,
-        #[cfg(all(hash_graph_test_environment, feature = "test-server"))]
-        Subcommand::TestServer(args) => subcommand::test_server(args).await,
-    }
+    let _sentry = sentry::init(sentry::ClientOptions {
```
Comment (TimDiekmann) [inline apps/hash-graph/bin/cli/src/main.rs:20]: For what the `_sentry` variable is used? If this is a guard we should name it like this and/or add a comment.

File: apps/hash-graph/bin/cli/src/subcommand.rs [hunk 5]
```diff
@@ -33,3 +36,30 @@ pub enum Subcommand {
...
+    }
+
+    pub(crate) fn execute(self) -> Result<(), GraphError> {
+        // create a runtime
+        let runtime = tokio::runtime::Builder::new_multi_thread()
+            .enable_all()
+            .build()
+            .expect("failed to create runtime");
+
+        runtime.block_on(self.delegate())
```
Comment (TimDiekmann) [inline apps/hash-graph/bin/cli/src/subcommand.rs:63]: Why do we block on `delegate` if not every subcommand in `delegate` is asynchronous? The `completions` subcommand does not require a runtime at all. I know that we previously had a similar behavior with `#[tokio::main]` but as we don't have this anymore we also don't need to spin up a runtime for that.

File: apps/hash-graph/bin/cli/src/main.rs [hunk 6]
```diff
@@ -5,27 +5,33 @@ mod args;
...
 
+use std::sync::Arc;
+
 use error_stack::Result;
 use graph::load_env;
 
-use self::{args::Args, error::GraphError, subcommand::Subcommand};
+use self::{args::Args, error::GraphError};
 
-#[tokio::main]
```
Comment (TimDiekmann) [inline apps/hash-graph/bin/cli/src/main.rs:13]: With this change it's possible to remove the `macros` feature from `tokio` I think.

File: apps/hash-realtime/src/index.ts [hunk 7]
```diff
@@ -66,8 +66,8 @@ const acquireReplicationSlot = async (
             -- Second argument to this calls specifies whether or not the replication
             -- slot is temporary.
             select 1 from pg_create_logical_replication_slot(${slotName}, 'wal2json', ${
-              options.temporary ?? true
-            })
+    options.temporary ?? true
+  })
```
Comment (TimDiekmann) [inline apps/hash-realtime/src/index.ts:70]: Do you have a different `prettier` version installed or why was this changed?

**Review by TimDiekmann — CHANGES_REQUESTED** [review]

File: apps/hash-graph/bin/cli/src/args.rs (hunk 2 above)
Comment (indietyp) [reply apps/hash-graph/bin/cli/src/args.rs]: I thought (from the clap documentation) that `Option<Option<T>>` would allow for empty values. I was wrong. Removed in [8e62433](https://github.com/hashintel/hash/pull/3091/commits/8e624331872d7bb6331f7e7bfd70a877aa182e66)

File: apps/hash-graph/bin/cli/src/main.rs (hunk 3 above)
Comment (indietyp) [reply apps/hash-graph/bin/cli/src/main.rs]: Nice catch! Fixed in [8e62433](https://github.com/hashintel/hash/pull/3091/commits/8e624331872d7bb6331f7e7bfd70a877aa182e66)

File: apps/hash-graph/bin/cli/src/main.rs (hunk 4 above)
Comment (indietyp) [reply apps/hash-graph/bin/cli/src/main.rs:26]: Added documentation in [8e62433](https://github.com/hashintel/hash/pull/3091/commits/8e624331872d7bb6331f7e7bfd70a877aa182e66)

File: apps/hash-graph/bin/cli/src/subcommand.rs (hunk 5 above)
Comment (indietyp) [reply apps/hash-graph/bin/cli/src/subcommand.rs]: made optional in [8e62433](https://github.com/hashintel/hash/pull/3091/commits/8e624331872d7bb6331f7e7bfd70a877aa182e66)

File: apps/hash-graph/bin/cli/src/main.rs (hunk 6 above)
Comment (indietyp) [reply apps/hash-graph/bin/cli/src/main.rs:13]: let's try it! removed in [c413faf](https://github.com/hashintel/hash/pull/3091/commits/c413faf59488886df56f0a4abc5162c3cf3412b3)

File: apps/hash-realtime/src/index.ts (hunk 7 above)
Comment (indietyp) [reply apps/hash-realtime/src/index.ts:70]: I was wondering about that myself, but I have yarn installed repeatedly (with `just yarn install`) and I always get the same result, 4 files in actuality changed. This one slipped through, the other three I correctly stashed.

File: apps/hash-graph/bin/cli/src/args.rs (hunk 2 above)
Comment (TimDiekmann) [reply apps/hash-graph/bin/cli/src/args.rs]: For what an empty value would be used?

Comment (indietyp) [reply apps/hash-graph/bin/cli/src/args.rs]: The idea was to set in the .env (and in the production `docker-compose.prod.yml`) the `HASH_GRAPH_SENTRY_DSN` to an empty value (so `HASH_GRAPH_SENTRY_DSN=`), which would then render to `None` because it is an empty value, so sentry isn't activated (I mean an empty string will stop startup anyway as it isn't a valid DSN)

File: apps/hash-realtime/src/index.ts (hunk 7 above)
Comment (indietyp) [reply apps/hash-realtime/src/index.ts:70]: Yea, if we look into the CI we can see:

```
Checking formatting...
[warn] apps/hash-realtime/src/index.ts
[warn] apps/hash/README.md
[warn] tests/hash-playwright/tests/shared/login-using-ui.ts
[warn] Code style issues found in 3 files. Run Prettier to fix.
```

Those are the other three files. Weird stuff.

Comment (TimDiekmann) [reply apps/hash-realtime/src/index.ts:70]: Well, it's good to have it fixed then I guess 😄

File: .env (hunk 1 above)
Comment (indietyp) [reply .env]: commented out in [8e62433](https://github.com/hashintel/hash/pull/3091/commits/8e624331872d7bb6331f7e7bfd70a877aa182e66)

File: apps/hash-graph/bin/cli/src/args.rs [hunk 8]
```diff
@@ -1,11 +1,16 @@
...
 
 /// Arguments passed to the program.
 #[derive(Debug, Parser)]
 #[clap(version, author, about, long_about = None)]
 pub struct Args {
+    // we need to qualify `Option` here, as otherwise `clap` tries to be too smart and only uses
+    // the `value_parser` on the internal `sentry::types::Dsn`, failing.
+    #[arg(long, env = "HASH_GRAPH_SENTRY_DSN", value_parser = OptionalSentryDsnParser)]
+    pub sentry_dsn: core::option::Option<sentry::types::Dsn>,
+
```
Comment (TimDiekmann) [inline apps/hash-graph/bin/cli/src/args.rs:13]: I wonder if this is supposed to be a global argument. However, the only service where it does not matter is the `completions` endpoint I'd assume. (non-blocking comment)

File: apps/hash-graph/bin/cli/src/main.rs [hunk 9]
```diff
@@ -3,29 +3,48 @@
...
-        Subcommand::TestServer(args) => subcommand::test_server(args).await,
+    let Args {
+        sentry_dsn,
+        subcommand,
+    } = Args::parse_args();
+
+    if let Some(dsn) = &sentry_dsn {
+        let client = sentry::Client::from_config(dsn);
+
+        assert!(client.is_enabled(), "Invalid Sentry DSN has been provided");
```
Comment (TimDiekmann) [inline apps/hash-graph/bin/cli/src/main.rs:27]: Do we want to return an error instead? I see why you use `assert!` as this is not recoverable at this point, however, it's directly next to `main()` and an error has the IMO better output.

**Review by TimDiekmann — COMMENTED** [review]
A few minor comments, looks good otherwise 🎉

File: apps/hash-graph/bin/cli/src/subcommand.rs [hunk 10]
```diff
@@ -33,3 +38,28 @@ pub enum Subcommand {
     #[cfg(all(hash_graph_test_environment, feature = "test-server"))]
     TestServer(TestServerArgs),
 }
+
+fn spawn_runtime(future: impl Future<Output = Result<(), GraphError>>) -> Result<(), GraphError> {
```
Comment (TimDiekmann) [inline apps/hash-graph/bin/cli/src/subcommand.rs:42]: Minor suggestion: Rename to `block_on`

File: apps/hash-graph/bin/cli/src/parser.rs [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+            let Some(value) = value.to_str() else {
+                return Err(Error::new(ErrorKind::InvalidValue).with_cmd(cmd));
+            };
+
+            sentry::types::Dsn::from_str(value)
+                .map(Some)
+                .map_err(|_error| Error::new(ErrorKind::InvalidValue))
+        }
+    }
+}
```
Comment (TimDiekmann) [inline apps/hash-graph/bin/cli/src/parser.rs:29]: For some reason the CI fails as it interprets this parameter as required.

Comment (indietyp) [reply apps/hash-graph/bin/cli/src/parser.rs:29]: This is because we use `core::option::Option<T>`, instead of `Option<T>`, which means that clap does not set it to `required = false`. We need to do this manually.

fixed in [6e591ab](https://github.com/hashintel/hash/pull/3091/commits/6e591ab5bb600a6b7881131b437fb0e044584f83)

File: apps/hash-graph/bin/cli/src/subcommand.rs (hunk 10 above)
Comment (indietyp) [reply apps/hash-graph/bin/cli/src/subcommand.rs]: renamed in [6e591ab](https://github.com/hashintel/hash/pull/3091/commits/6e591ab5bb600a6b7881131b437fb0e044584f83)

File: apps/hash-graph/bin/cli/src/main.rs (hunk 9 above)
Comment (indietyp) [reply apps/hash-graph/bin/cli/src/main.rs]: fixed in [6e591ab](https://github.com/hashintel/hash/pull/3091/commits/6e591ab5bb600a6b7881131b437fb0e044584f83)

File: apps/hash-external-services/docker-compose.prod.yml [hunk 12]
```diff
@@ -8,6 +8,7 @@ services:
   graph:
     image: hash-graph:prod
     environment:
+      HASH_GRAPH_SENTRY_DSN: "${HASH_GRAPH_SENTRY_DSN-}"
```
[the remark this answers is not in the document]
Comment (indietyp) [reply apps/hash-external-services/docker-compose.prod.yml]: fixed in [6e591ab](https://github.com/hashintel/hash/pull/3091/commits/6e591ab5bb600a6b7881131b437fb0e044584f83)

File: apps/hash-graph/bin/cli/src/args.rs (hunk 8 above)
Comment (indietyp) [reply apps/hash-graph/bin/cli/src/args.rs]: Yes, you are right, but (as you noted), I have decided to move it to a global flag, as the only command that does not benefit is the completions command, and it would lead to more code duplication (and make the guard more complicated)

> In my mind it also makes sense when you `--help`, as something like sentry or tracing should likely be a global argument and come before any subcommand. That way it is easier to switch out commands, etc.

Comment (TimDiekmann) [reply apps/hash-graph/bin/cli/src/args.rs]: That's fair. Do you know if it's possible to exclude this argument for certain subcommands? (Just wondering, doesn't block this PR)

> something like sentry or tracing should likely be a global argument and come before any subcommand

It's possible to achieve this with `clap` when making this a `global` argument.

File: apps/hash-graph/bin/cli/src/main.rs [hunk 13]
```diff
@@ -3,29 +3,51 @@
...
 
-    match args.subcommand {
-        Subcommand::Server(args) => subcommand::server(args).await,
-        Subcommand::Migrate(args) => subcommand::migrate(args).await,
-        Subcommand::TypeFetcher(args) => subcommand::type_fetcher(args).await,
-        Subcommand::Completions(ref args) => {
-            subcommand::completions(args);
-            Ok(())
+        if !client.is_enabled() {
+            return Err(Report::new(SentryError::InvalidDsn).change_context(GraphError));
```
Comment (TimDiekmann) [inline apps/hash-graph/bin/cli/src/main.rs:29]: Let's attach the DSN. These kinds of attachments makes life so much easier with `error-stack` and this should be easy enough I think.

**Review by TimDiekmann — DISMISSED** [review]

File: apps/hash-graph/bin/cli/src/main.rs (hunk 13 above)
Comment (indietyp) [reply apps/hash-graph/bin/cli/src/main.rs]: Good call! ([a5b4bcb](https://github.com/hashintel/hash/pull/3091/commits/a5b4bcbb75caaf586b4c96766d4ce6d094380973))

File: apps/hash-graph/bin/cli/src/main.rs [hunk 14]
```diff
@@ -3,29 +3,53 @@
...
-        Subcommand::Migrate(args) => subcommand::migrate(args).await,
-        Subcommand::TypeFetcher(args) => subcommand::type_fetcher(args).await,
-        Subcommand::Completions(ref args) => {
-            subcommand::completions(args);
-            Ok(())
+        if !client.is_enabled() {
+            return Err(Report::new(SentryError::InvalidDsn)
+                .attach_printable(dsn.clone())
+                .change_context(GraphError));
         }
```
Comment (TimDiekmann) [inline apps/hash-graph/bin/cli/src/main.rs:32]:
```suggestion
        ensure!(
            client.is_enabled(),
            Report::new(SentryError::InvalidDsn)
                .attach_printable(dsn.clone())
                .change_context(GraphError)
        );
```

**Review by TimDiekmann — APPROVED** [review]
Whatever style you prefer 😉

**Review by TimDiekmann — APPROVED** [review]

**Review by TimDiekmann — APPROVED** [review]
